src/dtls/network.c

```diff
@@ -0,0 +1,962 @@
+}
+
+
+//
+// DATA TRANSMISSION CALLBACKS
+//
+// mbedtls uses callbacks to read data from the network and to write data for
+// the network. Note that when our code gets new data from the network, it needs
+// to trigger mbedtls that will then callback to read the data. Unfortunately
+// mbedtls doesn't provide a way to directly feed the data into its state
+// machine.
+//
+// OnTLSRecv() consumes the read queue. OnTLSSend uses uv_udp functions to write
+// to the network directly.
+//
+
+static int OnTLSRecv(void* data, unsigned char *buf, size_t len)
+{
+    DPS_DBGTRACE();
+
+    DPS_NetConnection* cn = data;
+    assert(IsNodeThread(cn->node));
+
+    DPS_DBGPRINT("OnTLSRecv() want to read using %zu bytes of buffer\n", len);
+
+    PendingRead* pr = cn->readQueue;
+    if (!pr) {
+        DPS_DBGPRINT("OnTLSRecv() pending read empty\n");
+        return MBEDTLS_ERR_SSL_WANT_READ;
+    }
+
+    DPS_DBGPRINT("OnTLSRecv() using pending read with %zu bytes\n", pr->buf.len);
+    cn->readQueue = pr->next;
+
+    if (pr->buf.len > len) {
+        return MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL;
+    }
+
+    size_t dataLen = pr->buf.len;
+    memcpy(buf, pr->buf.base, dataLen);
+
+    free(pr->buf.base);
+    free(pr);
+    return dataLen;
+}
+
+static void OnTLSSendComplete(uv_udp_send_t *req, int status)
+{
+    DPS_DBGPRINT("OnTLSSendComplete cleaning up for send req=%p\n", req);
+    if (status != 0) {
+        DPS_ERRPRINT("ERROR: couldn't send asynchronously with status=%d: %s\n", status, uv_err_name(status));
+        // TODO: Add DPS_SendFailed() call.
+    }
+    free(req->data);
+    free(req);
+}
+
+static int OnTLSSend(void* data, const unsigned char *buf, size_t len)
+{
+    DPS_NetConnection* cn = data;
+    assert(IsNodeThread(cn->node));
+
+    DPS_DBGPRINT("OnTLSSend want to write %zu bytes\n", len);
+
+    uv_udp_send_t* req = calloc(1, sizeof(uv_udp_send_t));
+    if (!req) {
+        return -1;
+    }
+
+    uv_buf_t ubuf = {};
+
+
+    // TODO: Verify the statement below and whether we couldn't own the buffer.
+
+    // We don't own the buffer mbedtls passed us, so we need to copy so that the
+    // UDP request used it later.
+    ubuf.base = malloc(len);
+    if (!ubuf.base) {
+        free(req);
+        return -1;
+    }
+
+    memcpy_s(ubuf.base, len, buf, len);
+    ubuf.len = len;
+    req->data = ubuf.base;
+
+    int err = uv_udp_send(req, cn->socket, &ubuf, 1, (const struct sockaddr *)&cn->peer->addr.inaddr, OnTLSSendComplete);
+    if (err) {
+        DPS_ERRPRINT("ERROR: couldn't send UDP packet: %s\n", uv_err_name(err));
+        return -1;
+    }
+
+    return len;
+}
+
+static void CancelPendingWrites(DPS_NetConnection* cn, PendingWrite* pw)
+{
+    while (pw) {
+        PendingWrite* next = pw->next;
+        pw->sendCompleteCB(cn->node, pw->appCtx, cn->peer, pw->bufs, pw->numBufs, DPS_ERR_NETWORK);
+        free(pw->bufs);
+        free(pw);
+        DPS_DBGPRINT("freeing PendingWrite %p", pw);
```

Comment:
Indeed not a problem, but I've shuffled these around, so the non-problem is not there anymore :-)